rllib/policy/sample_batch.py

```diff
@@ -133,23 +133,43 @@ def __init__(self, *args, **kwargs):
         if self._is_training is None:
             self._is_training = self.pop("is_training", False)
 
+        # Try to infer the "length" of the SampleBatch by finding the first
+        # value that is actually a ndarray/tensor.
         lengths = []
         copy_ = {k: v for k, v in self.items() if k != SampleBatch.SEQ_LENS}
         for k, v in copy_.items():
             assert isinstance(k, str), self
 
+            # If this is a nested dict (for example a nested observation),
+            # try to flatten it, assert that all elements have the same length (batch
+            # dimension)
+            if isinstance(v, (dict, tuple)):
+                if k == SampleBatch.INFOS:
+                    # Don't attempt to count on infos since we make no assumptions
+                    # about its content
+                    continue
+                v = tree.flatten(v)
+                try:
+                    # Add one of the elements length, since they are all the same
+                    len_ = len(v[0])
+                    if len_:
+                        lengths.append(len_)
```

Comment:
remove if here:
```
lengths.append(len(v[0]))
```

Reply:
This would change the logic of the whole function because we further down depend on lengths[0].
Right now, `lengths` can never contain zeros.

Reply:
Don't get me wrong - I want this and did it already before submitting.
But this would enlargen the scope of what can happen here significantly and even break user code.

Reply:
oh so len_ should not be appended if it's zero. I see. makes sense.